### ISSUE TITLE
Another asSuperVisitor crash

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#2370]
I believe I've encountered another issue with `asSuperVisitor`, with superficial similarities to those fixed in #2193. However, this crash occurs on the current master version of the framework.

Consider the following test case:

```java
import java.util.stream.Stream;
import java.util.Collection;
import java.util.Optional;

public class StreamCrash {
    private Stream<Action> getActions(final State state) {
	return Stream.of(
			 toStream(state.getOnExit()).flatMap(t -> t.getActions().stream()),
			 toStream(state.getOnSignal()).flatMap(t -> t.getActions().stream()),
			 toStream(state.getOnEnter()).flatMap(t -> t.getActions().stream())
			 ).flatMap(actionStream -> actionStream);
    }
    
    private <T> Stream<T> toStream(final Collection<T> obj) {
        return Optional.ofNullable(obj)
                .map(Stream::of)
                .orElseGet(Stream::empty)
                .flatMap(Collection::stream);
    }
}

interface Action {
    public Collection<Action> getActions();
}

interface State {
    public Collection<Action> getOnExit();
    public Collection<Action> getOnSignal();
    public Collection<Action> getOnEnter();
}
```

When running any typechecker (tested with the Value Checker and Nullness Checker) on this file, a crash occurs in `asSuperVisitor`. The command to run is `> checker/bin-devel/javac -processor org.checkerframework.common.value.ValueChecker StreamCrash.java`. The resulting stack trace looks like this:

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @UnknownVal Object
  superType: @UnknownVal Action
  Compilation unit: StreamCrash.java
  Last visited tree at line 5 column 1:
  public class StreamCrash {
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:25)
  org.checkerframework.framework.type.AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:154)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:350)
  org.checkerframework.framework.type.AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:28)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.visitTypevar_NotTypevarNorWildcard(AsSuperVisitor.java:595)
  org.checkerframework.framework.type.AsSuperVisitor.visitTypevar_Declared(AsSuperVisitor.java:602)
  org.checkerframework.framework.type.AsSuperVisitor.visitTypevar_Declared(AsSuperVisitor.java:28)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:631)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:57)
  org.checkerframework.framework.type.AsSuperVisitor.visit(AsSuperVisitor.java:85)
  org.checkerframework.framework.type.AsSuperVisitor.asSuper(AsSuperVisitor.java:67)
  org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:406)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:472)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:290)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.scan(EquivalentAtmComboScanner.java:56)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.visit(EquivalentAtmComboScanner.java:36)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:408)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:472)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visitDeclared_Declared(AbstractQualifierPolymorphism.java:290)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.scan(EquivalentAtmComboScanner.java:56)
  org.checkerframework.framework.type.visitor.EquivalentAtmComboScanner.visit(EquivalentAtmComboScanner.java:36)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.visit(AbstractQualifierPolymorphism.java:408)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism$PolyCollector.access$200(AbstractQualifierPolymorphism.java:290)
  org.checkerframework.framework.type.poly.AbstractQualifierPolymorphism.annotate(AbstractQualifierPolymorphism.java:147)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1571)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:170)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:44)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1272)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1052)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:201)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:925)
  org.checkerframework.common.value.ValueTransfer.visitMethodInvocation(ValueTransfer.java:385)
  org.checkerframework.common.value.ValueTransfer.visitMethodInvocation(ValueTransfer.java:68)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:76)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:408)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysisBlock(Analysis.java:234)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:187)
  org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:91)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1228)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1146)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1518)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:260)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:320)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:980)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:490)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
1 error
exit 1
```

**Comment by martinschaef** [comment]
Thx @kelloggm for filing this. We ran into this on two occasions, and it is likely to happen more frequently, since we make heavy use of streams.

**Comment by smillst** [comment]
The fixes in https://github.com/typetools/checker-framework/pull/2193 were not correct.  I've reverted some of those changes and added the correct fixes in PR #2374.  Those fixes also fix this crash. 

FYI, "AsSuper" crashes like this one mean that a caller to asSuper is violating asSupers preconditions.  (Namely that the underlying java types passed to supertype is a supertype of some type.)  So anytime you see a crash like this with a different stack after ` org.checkerframework.framework.util.AnnotatedTypes.asSuper(AnnotatedTypes.java:112)`, it is potentially a distinct crash.
